trieste/models/gpflow/models.py

```diff
@@ -507,13 +496,291 @@ def conditional_predict_y(
+        # GPflow stores the data in Tensors. However, since we want to be able to update the data
+        # without having to retrace the acquisition functions, put it in Variables instead.
+        # Data has to be stored in variables with dynamic shape to allow for changes
+        # Sometimes, for instance after serialization-deserialization, the shape can be overridden
+        # Thus here we ensure data is stored in dynamic shape Variables
+        if not all(is_variable(x) and x.shape[0] is None for x in self._model.data):
+            self._model.data = (
+                tf.Variable(
+                    self._model.data[0],
+                    trainable=False,
+                    shape=[None, *self._model.data[0].shape[1:]],
+                ),
+                tf.Variable(
+                    self._model.data[1],
+                    trainable=False,
+                    shape=[None, *self._model.data[1].shape[1:]],
+                ),
+            )
+
+        if not is_variable(self._model.num_data):
+            self._model.num_data = tf.Variable(self._model.num_data, trainable=False)
+
+    def optimize(self, dataset: Dataset) -> None:
+        """
+        Optimize the model with the specified `dataset`.
+
+        :param dataset: The data with which to optimize the `model`.
+        """
+        self.optimizer.optimize(self.model, dataset)
+        self.update_posterior_cache()
+
+    def update(self, dataset: Dataset) -> None:
+        self._ensure_variable_model_data()
+
+        x, y = self.model.data[0].value(), self.model.data[1].value()
+
+        assert_data_is_compatible(dataset, Dataset(x, y))
+
+        if dataset.query_points.shape[-1] != x.shape[-1]:
+            raise ValueError
+
+        if dataset.observations.shape[-1] != y.shape[-1]:
+            raise ValueError
+
+        self.model.data[0].assign(dataset.query_points)
+        self.model.data[1].assign(dataset.observations)
+
+        current_inducing_points, q_mu, _, _ = self.get_inducing_variables()
+
+        if isinstance(current_inducing_points, list):
+            inducing_points_trailing_dim = current_inducing_points[0].shape[-1]
+        else:
+            inducing_points_trailing_dim = current_inducing_points.shape[-1]
+
+        if dataset.query_points.shape[-1] != inducing_points_trailing_dim:
+            raise ValueError(
+                f"Shape {dataset.query_points.shape} of new query points is incompatible with"
+                f" shape {self.model.inducing_variable.Z.shape} of existing query points."
+                f" Trailing dimensions must match."
+            )
+
+        if dataset.observations.shape[-1] != q_mu.shape[-1]:
+            raise ValueError(
+                f"Shape {dataset.observations.shape} of new observations is incompatible with"
+                f" shape {self.model.q_mu.shape} of existing observations. Trailing"
+                f" dimensions must match."
+            )
```

Comment:
@henrymoss I have copied this part from SVGP, but I suspect this was needed there as data is not an argument of the model and provided only at optimisation and update time?

since we already do a check of the data above and q_mu is computed with respect to that data we should be fine without this check?

Reply:
This is @uri-granta 's area?

Reply:
It is?!

My vote would be to leave it in, as it's still a valid and cheap check on the new observations.